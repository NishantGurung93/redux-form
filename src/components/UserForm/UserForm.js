import { Form, Formik } from "formik";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router-dom";
import { submitDetails } from "../../slices/userSlice";
import { validateForm } from "../../utils/validations";
import Input from "../Input";

const UserForm = () => {
  const dispatch = useDispatch();
  const history = useHistory();

  const initialValues = {
    name: '',
    role: '',
    email: '',
    password: '',
  };

  return (
    <Formik
      initialValues={initialValues}
      validate={validateForm}
      onSubmit={(values) => {
        dispatch(submitDetails(values));
        history.push('/privacy');
      }}
    >

      <Form>
        <Input
          label="Name"
          name="name"
          placeholder="Ash"
          type="text"
          required
        />

        <Input
          label="Role"
          name="role"
          placeholder="Software engineer"
          type="text"
        />

        <Input
          label="Email"
          name="email"
          placeholder="ash@ketchum.com"
          type="email"
          required
        />

        <Input
          label="Password"
          name="password"
          placeholder="Password"
          type="password"
          required
        />

        <button type="submit" data-testid="submit-button">Submit</button>
      </Form>
    </Formik>
  )
};

export default UserForm;
