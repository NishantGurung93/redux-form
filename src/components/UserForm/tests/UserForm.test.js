import { fireEvent, render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import UserForm from "..";

const dispatch = jest.fn();
const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };

const mockStore = {
  value: {},
  useDispatch: jest.fn(),
  dispatch,
  getState: jest.fn(),
  subscribe: jest.fn(),
}

const renderComponent = () => render(
  <Provider store={mockStore}>
    <Router history={historyMock}>
      <UserForm />
    </Router>
  </Provider>
);

describe('Header component', () => {
  it('should render the component', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  describe('when submitting a valid form', () => {
    beforeEach(async () => {
      const { getByTestId } = renderComponent();

      const name = getByTestId('name-input');
      fireEvent.change(name, { target: { value: 'Tom Hardy' } });

      const role = getByTestId('role-input');
      fireEvent.change(role, { target: { value: 'Bane' } });

      const email = getByTestId('email-input');
      fireEvent.change(email, { target: { value: 'tom@hardy.com' } });

      const password = getByTestId('password-input');
      fireEvent.change(password, { target: { value: 'SuperCoolTom123' } });

      const submitButton = getByTestId('submit-button');

      await act(async () => {
        fireEvent.click(submitButton);
      });
    });

    it('should call submitDetails', () => {
      expect(dispatch).toHaveBeenCalledWith({
        payload: {
          email: 'tom@hardy.com',
          name: 'Tom Hardy',
          password: 'SuperCoolTom123',
          role: 'Bane',
        },
        type: 'user/submitDetails',
      });
    });

    it('should call useHistory hook with "/privacy"', () => {
      expect(historyMock.push).toHaveBeenCalledWith('/privacy');
    });
  });

  describe('when submitting an invalid form', () => {
    beforeEach(async () => {
      const { getByTestId } = renderComponent();

      const submitButton = getByTestId('submit-button');

      await act(async () => {
        fireEvent.click(submitButton);
      });
    });

    it('should not call submitDetails', () => {
      expect(dispatch).not.toHaveBeenCalled();
    });

    it('should not call useHistory hook', () => {
      expect(historyMock.push).not.toHaveBeenCalled();
    });
  });
});
