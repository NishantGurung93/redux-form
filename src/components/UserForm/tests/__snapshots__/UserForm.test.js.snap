// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render the component 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="input-wrapper"
    >
      <label
        for="name"
      >
        Name:
        <span>
           *
        </span>
      </label>
      <input
        data-testid="name-input"
        id="name"
        name="name"
        placeholder="Ash"
        type="text"
        value=""
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        for="role"
      >
        Role:
      </label>
      <input
        data-testid="role-input"
        id="role"
        name="role"
        placeholder="Software engineer"
        type="text"
        value=""
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        for="email"
      >
        Email:
        <span>
           *
        </span>
      </label>
      <input
        data-testid="email-input"
        id="email"
        name="email"
        placeholder="ash@ketchum.com"
        type="email"
        value=""
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        for="password"
      >
        Password:
        <span>
           *
        </span>
      </label>
      <input
        data-testid="password-input"
        id="password"
        name="password"
        placeholder="**********"
        type="password"
        value=""
      />
    </div>
    <button
      class="submit-button"
      data-testid="submit-button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
